src/ripple/overlay/impl/PeerSet.cpp

```diff
@@ -46,15 +46,41 @@ PeerSet::PeerSet(
 
 PeerSet::~PeerSet() = default;
 
-bool PeerSet::insert (std::shared_ptr<Peer> const& ptr)
+void
+PeerSet::addPeers(
+    std::size_t limit,
+    std::function<bool(std::shared_ptr<Peer> const&)> hasItem)
 {
-    ScopedLockType sl (mLock);
+    using ScoredPeer = std::pair<int, std::shared_ptr<Peer>>;
+
+    auto const& overlay = app_.overlay();
 
-    if (!mPeers.insert (ptr->id ()).second)
-        return false;
+    std::vector<ScoredPeer> pairs;
+    pairs.reserve(overlay.size());
 
-    newPeer (ptr);
-    return true;
+    overlay.foreach([&](auto const& peer) {
+        auto const score = peer->getScore(hasItem(peer));
+        pairs.emplace_back(score, std::move(peer));
+    });
+
+    std::sort(
+        pairs.begin(),
+        pairs.end(),
+        [](ScoredPeer const& lhs, ScoredPeer const& rhs) {
+            return lhs.first > rhs.first;
+        });
+
+    std::size_t accepted = 0;
+    ScopedLockType sl(mLock);
+    for (auto const& pair : pairs)
+    {
+        auto const peer = pair.second;
+        if (!mPeers.insert(peer->id()).second)
+            continue;
+        newPeer(peer);
```

Comment:
Is "this" referring to the call to `newPeer()`? The old implementation of `selectPeers()` [called](https://github.com/ripple/rippled/blob/0c6d380780ae368a2236a2e8e3e42efa4a1d2b46/src/ripple/overlay/impl/OverlayImpl.cpp#L873) `PeerSet::insert` (not `std::set::insert` directly), [which](https://github.com/ripple/rippled/blob/4a1148eb2849513dd1e7ae080288fd47ab57a376/src/ripple/overlay/impl/PeerSet.cpp#L58-L67) acquired the lock and called `newPeer()`. Instead of toggling the lock on and off, I figured I'd move it out of the loop and acquire it once for the whole set.